design.md

```diff
@@ -0,0 +1,99 @@
+- a base workflow engine ([Cromwell](vidarr-cromwell/))
+
+Additional components include:
+
+- [Tools](https://github.com/oicr-gsi/vidarr-tools/) to generate workflow definitions
+- [Cerberus](http://github.com/oicr-gsi/cerberus) create file provenance by integrating Víðarr output with Pinery
+
+The goal for Víðarr web service is to:
+
+- Take workflow requests.
+- Check if this request is already completed.
+- Delegate the workflow to a workflow engine.
+- Provision the output and record the output in a database.
+- Provide an interface to this database similar to provide a history of completed workflows.
+
+## Shesmu Integration
+Shesmu needs file provenance and can use Cerberus as a plugin. Shesmu also
+needs run workflows and can use the `/api/submit` endpoint for that purpose. It
+needs to know what it can run and uses `/api/workflows` to get the known
+workflow and `/api/targets` to know where it can run them; it generates an
+action definition for every valid workflow version and target combination. The
+submit action in Shesmu has commands to re-run failed workflows (_Reattempt_)
+and unload-and-rerun (_Reprocess_). For more details, see the [Víðarr plugin
+for
+Shesmu](https://github.com/oicr-gsi/shesmu/blob/master/plugin-vidarr/README.md).
+
+## How a workflow runs
+Víðarr takes a submission request that includes:
+
+- the workflow name and version
+- the target name
+- the parameters to the workflows
+- the external keys (LIMS keys) associated with the workflow
+- the output metadata, that indicates what output is attached to which output
+- any labels required by the workflow
+- any parameters to the workflow engine
+- consumable resource values
+
+Once received:
+
+1. The target and workflow name and version are checked. If not found or the workflow is not supported by the workflow engine, it is rejected.
+2. The parameters, metadata, labels, consumable resource values, and engine parameters are type checked.
+3. References to other Víðarr files used as input are resolved.
+4. External keys are checked.
+5. The [identifier](identifiers.md) is calculated.
+6. The database is checked for the identifier.
+   If it is found and in a failed state and the request asked for reattempting, the workflow is started.
+   If it is found, the identifier is returned to the client.
+   If the identifier is not found, the workflow run is started.
+
+Once a workflow is started (or restarted), it goes through phases:
+
+- the workflow waits for any consumable resource to allow it to continue
+- the workflow does a preflight check where it ensure that configuration in the output metadata is valid
+- it provisions in any input files
+- it runs the workflow
+- it provisions out the results of the workflow
+- it performs any clean up required
+
+We keep running into issue of running out of temp. Shesmu can throttle, but
+it's often too late. The Víðarr has a number of consumable resources (_e.g._,
```

Comment:
The Vidarr -> Vidarr? Or the Vidarr server?

Reply:
I think Vidarr server makes more sense.